src/HPLTokenHandler.mo

```diff
@@ -0,0 +1,405 @@
+
+  public func defaultHandlerStableData(): StableData = ([], 0, (0, 0), ([var], 0, 0));
+
+  public type Info = {
+    var credit : Nat;
+    virtualAccountId : HPL.VirtualAccountId;
+  };
+
+  public type StableInfo = {
+    credit : Nat;
+    virtualAccountId : HPL.VirtualAccountId;
+  };
+
+  public type JournalRecord = (Time.Time, Principal, {
+    #credited: Nat;
+    #debited: Nat;
+    #error: Any;
+    #openAccountError: { #UnknownPrincipal; #UnknownSubaccount; #MismatchInAsset; #NoSpaceForAccount };
+    #sweepIn: Nat;
+    #sweepOut: Nat;
+    #withdraw: { to: (Principal, HPL.AccountReference); amount: Nat };
+    #deposit: { from: (Principal, HPL.VirtualAccountId); amount: Nat };
+  });
+
+  public type DepositError = {
+    #CallHPLError;
+    #DeletedVirtualAccount;
+    #InsufficientFunds;
+    #MismatchInAsset;
+    #MismatchInRemotePrincipal; 
+    #TooLargeFtQuantity;
+    #TooLargeVirtualAccountId;
+    #UnknownPrincipal;
+    #UnknownVirtualAccount;
+  };
+
+  public type WithdrawError = {
+    #CallHPLError;
+    #DeletedVirtualAccount;
+    #MismatchInAsset;
+    #MismatchInRemotePrincipal; 
+    #TooLargeFtQuantity;
+    #TooLargeSubaccountId;
+    #TooLargeVirtualAccountId;
+    #UnknownPrincipal;
+    #UnknownSubaccount;
+    #UnknownVirtualAccount;
+  };
+
+  public type StableData = (
+    [(Principal, StableInfo)],        // map
+    Nat,                              // consolidatedFunds
+    (Nat, Nat),                       // total debited/credited
+    ([var ?JournalRecord], Nat, Int)  // journal
+  );
+
+  public class TokenHandler(
+    hplLedgerPrincipal : Principal,
+    assetId: HPL.AssetId, 
+    backingSubaccountId: HPL.SubaccountId,
+    ownPrincipal : Principal,
+    journalSize: Nat,
+  ) {
+    let hpl = actor (Principal.toText(hplLedgerPrincipal)) : HPL.Ledger;
+
+    /// if some unexpected error happened, this flag turns true and handler stops doing anything until recreated
+    var isFrozen_ : Bool = false;
+    func freezeTokenHandler(errorText: Text): () {
+      isFrozen_ := true;
+      journal.push((Time.now(), ownPrincipal, #error(errorText)));
+    };
+
+    var journal : CircularBuffer.CircularBuffer<JournalRecord> = CircularBuffer.CircularBuffer<JournalRecord>(journalSize);
+    var consolidatedFunds_ : Nat = 0;
+    var map : RBTree.RBTree<Principal, Info> = RBTree.RBTree<Principal, Info>(Principal.compare);
+
+    var totalDebited : Nat = 0;
+    var totalCredited : Nat = 0;
+
+    /// Returns reference to registered virtual account for P.
+    /// If not registered yet, registers it automatically
+    public func getAccountReferenceFor(p : Principal) : async* (Principal, HPL.VirtualAccountId) {
+      let virtualAccountId : HPL.VirtualAccountId = switch (map.get(p)) {
+        case (?info) info.virtualAccountId;
+        case (null) {
+          let registerResult = await hpl.openVirtualAccount({ asset = #ft(assetId, 0); backingSubaccountId = backingSubaccountId; remotePrincipal = p });
```

Comment:
We don't have try-catch here. Is that intended? It means a call error gets propagated back to the enclosing async. Maybe that is intended?

Reply:
Propagating is fine. But we should add a comment:
```
// We pass through any call Error instead of catching it
```

Reply:
Added logic to handle call errors, as we do in other places